File: Difficulty: Medium/Longest Bounded-Difference Subarray/longest-bounded-difference-subarray.py
class Solution:
    def longestSubarray(self, arr, x):
        if not arr:
            return []
        
        n = len(arr)
        if n == 1:
            return arr
            
        # Initialize window parameters
        max_length = 1
        start_idx = 0
        window_min = arr[0]
        window_max = arr[0]
        left = 0
        
        # Single pass with sliding window
        for right in range(1, n):
            # Update window min and max
            window_min = min(window_min, arr[right])
            window_max = max(window_max, arr[right])
            
            # Check if current window is valid
            while window_max - window_min > x:
                # Shrink window from left
                left += 1
                # Update min and max based on remaining elements
                window_min = min(arr[left:right + 1])
                window_max = max(arr[left:right + 1])
            
            # Update result if current window is longer
            curr_length = right - left + 1
            if curr_length > max_length:
                max_length = curr_length
                start_idx = left
        
        return arr[start_idx:start_idx + max_length]


#{ 
 # Driver Code Starts
#Initial Template for Python
if __name__ == '__main__':
    tc = int(input())
    while tc > 0:
        arr = list(map(int, input().strip().split()))
        k = int(input().strip())
        ob = Solution()
        ans = ob.longestSubarray(arr, k)

        print(" ".join(map(str, ans)))
        tc -= 1
        print("~")

# } Driver Code Ends

File: Difficulty: Medium/Longest Bounded-Difference Subarray/longest-bounded-difference-subarray.java
import java.util.*;

class Solution {
    public ArrayList<Integer> longestSubarray(int[] arr, int x) {
        int n = arr.length;
        int left = 0, right = 0;
        int maxLen = 0, startIdx = 0; // Store length and start index of longest valid subarray

        TreeMap<Integer, Integer> map = new TreeMap<>();
        
        while (right < n) {
            // Add current element to the TreeMap
            map.put(arr[right], map.getOrDefault(arr[right], 0) + 1);

            // Check window's min and max; shrink window if needed
            while (map.lastKey() - map.firstKey() > x) {
                map.put(arr[left], map.get(arr[left]) - 1);
                if (map.get(arr[left]) == 0) {
                    map.remove(arr[left]);
                }
                left++;
            }

            // Update result if current window is longer
            if (right - left + 1 > maxLen) {
                maxLen = right - left + 1;
                startIdx = left;
            }
            right++;
        }

        // Prepare output: longest valid subarray starting earliest
        ArrayList<Integer> ans = new ArrayList<>();
        for (int i = startIdx; i < startIdx + maxLen; i++) {
            ans.add(arr[i]);
        }
        return ans;
    }
}
